### ISSUE TITLE
Thoughts on the evolution of the code base

### ISSUE THREAD

**Issue by rixner** [skulpt/skulpt#867]
While working on #861, I came to two conclusions:

1. The closure library is probably not pulling it's weight.
2. I think ```skulpt.py``` should be rewritten in Javascript.

First, at this point, from the closure library, we are using ```goog.exportSymbol```, ```goog.asserts.assert```, ```goog.asserts.fail```, and a few various ```goog.string``` functions.  Presumably, we could write our own versions of these few functions (or use lighter weight packages that do the same thing).  Instead, we're adding some 4000 lines of Javascript into our distribution from a very old version of the closure library.  While the closure compiler still probably has some value, I really don't think the closure library does for us.  Furthermore, the closure library does not appear to be modularized in a modern way, so if we start rewriting the core of Skulpt, it will take some effort to pull in just the pieces of the closure library we want.

Second, I think it will be easier overall to use node to manage and run things if the control code is in Javascript instead of Python.  This is likely something we could do incrementally, moving pieces of the infrastructure to ```skulpt.js``` until ```skulpt.py``` is obsolete.

Just my thoughts.  Figured I'd record them here for posterity.

**Comment by rixner** [comment]
I have successfully removed all use of the closure library from the code.  I will make a pull request after #861 is merged.

**Comment by rixner** [comment]
I now also am able to build Skulpt with webpack.  And I can run the closure compiler via webpack.  I also have it run jshint during the build.  I cannot get jscs to work correctly with the current version of webpack, but I think that's largely because jscs has been subsumed by eslint.  Eslint does run, but with the automatically converted rc file I made it throws tons of errors.  I think it would be good for us to convert to using eslint, but we need to agree on an rc file (preferably one that enforces our existing style without adding new restrictions, in my opinion).

In converting to webpack, it also obsoletes a lot of things in skulpt.py, so that's getting smaller and smaller.  I do have a few questions for people:

1. What's with the STANDARD_NAMING if statements?  Do we really need to carry the non-standard names, or can I just drop them entirely?
2. I favor always building either a compressed or uncompressed skulpt.min.js/skulpt.js file before running or testing anything.  That ensures we are testing an actual build with everything defined properly, rather than just importing the entry point and hoping that's the same.  Are people okay with that?
3. If the answer to (2) is "yes", then should build and run/test be two different steps?  Or should running something always do a build first?  The latter is slower, but it ensures you don't use a stale build.  Presumably, someone more webpack-savvy than I could eventually incorporate watching to automatically rebuild on source code changes.  I favor the former, but I can see both sides.
4. I don't use a lot of the browser stuff and a lot of it broke when I moved things to node.  I think that could be rewritten to be much cleaner using express by someone who actually uses it and knows what they want from it.  And it can/should use the same built skulpt.js/skulpt.min.js files as everything else.  Are people okay if I don't fix this and leave it to someone else?
5. There's a lot of test building stuff that I think is largely dead (regentests, regensymtabtests, regenasttests, regenruntests, ...).  I am inclined to just drop those commands entirely.  We have the output in the test directories, so the tests can all still be run.  But, they are all Python 2 things that we haven't regenerated in 5+ years.  Moving forward, all new tests are unit tests, so I don't see the utility of these other than for historical reasons, and they'll be in the git history for that...  These seem to be the only things that really have any dependence on Python.
6. I have been creating individual .js files for each command (run, test, etc.) and then running that file through node via package.json.  Do people prefer the monolithic skulpt.js approach?  There's really not as much shared once we go this route, so I didn't see the value.  But, will do what others think is best.  As an example, to run the repl, all that is really needed is to run ```node repl/repl.js```.  It seems like having ```"repl": "node repl/repl.js"``` in package.json makes much more sense than ```"repl": "node skulpt.js repl``` and then parsing that command and firing off the repl.  Most commands are similarly simple, because you don't need to construct the javacript on the fly, which is what skulpt.py was doing.  So, the ```run``` command, for example can just be ```node test/runfile.js <filename>``` where ```runfile.js``` is basically just the premade contents that skulpt.py was generating.  The other cruft around it becomes unnecessary.

**Comment by bnmnetp** [comment]
1. -- maybe lost to the annals of time.. But I don't see why we would need non-standard names anymore.

2. I'm in favor

3. I almost always do a `./skulpt.py dist` So I'm happy to have the two things combined to save me the keystrokes.  I would not want to lose the ability to do something equivalent to `./skulpt.py run xx.py` for testing and fixing bugs.

4. I find the browserdebug stuff to be really valueable.  But I haven't done it for a while.  

5. I agree that all of these old things can be removed...  I've been wondering about the utility of all of those test/run/* files I think that the vast majority (maybe all?) of them are redundant given our unittests now.  But that can be a separate discussion.

6. I'm not very advanced at writing javascript apps for the command line, so I don't have any strong opinions on this.  Is there some kind of best practice recognized by the JS community that we should just follow?

**Comment by albertjan** [comment]
1. I don't know what you mean with that 😄 not worried
2. `run`, `run3`, `brun` and `brun3` are my main avenues for testing things I really can't live without them. They should include all the files seperately (like it does now) or one file with a source map with the separate files for debugging in the browser. These should **not** rely on a full build, with tests and the closure compiler. If I have to wait for all the tests and the compiler after every change I do, that would really slow me down. 
3. see my answer on 2. 😄 
4. I don't use browerdebug nor do I run the test suite in the browser anymore. I run `./m test` to find the broken test and then run `./m brun path/to/file.py` to debug what happens in the browser. I was about to submit a tiny PR to fix `brun` 😄 
5. `(regentests, regensymtabtests, regenasttests, regenruntests)` I've used them, although we've disabled the ast and symtab tests. I for one am sure that @meredydd will want to resurrect them at some point, at least the ast tests.
6. I've always had this idea to have like a `makefile` like thing that can run tests and build skulpt and such. But that we also have an entry point that if you build skulpt you'll end up with a `skulpt` executable (like thing) that when executed would start the repl or when you say `./skulpt test.py` it would run the file. 

I'd like a repl3 but that seems to always break for me 😄

**Comment by albertjan** [comment]
About `debugbrowser` that can go, it's old and not really handy to work with. It's very finicky on windows with the line endings and the diffing of the run results. I'm for removing this and redoing it properly integrating with a browser testing framework.

**Comment by bnmnetp** [comment]
I meant `brun` not browserdebug. 😄

**Comment by rixner** [comment]
To give you an idea of what "building" means, I created a ```devbuild``` and ```build``` script/command.  The first just has webpack create an uncompressed ```skulpt.js``` with a source map.  The latter runs the closure compiler to build a compressed ```skulpt.min.js```, also with a source map.  They both run jshint and eslint.  Neither command runs any tests.

[build.txt](https://github.com/skulpt/skulpt/files/3213124/build.txt)

@albertjan is this too slow for your workflow?

@bnmnetp I can easily add a ```dist``` command that does this build first and then runs the tests.  I will point out that the current dist command does *not* run the Python 3 tests, which are currently broken because they get ignored.  I would like to fix them and then force them to run in Travis before we accept pull requests.

The question I have for both the ```run``` command (which will still exist) and @albertjan 's workflow, is whether you want to use the unbundled files (which would be faster, as you don't have to wait for the build) or not.  If you are always going to rebuild between runs, then maybe requiring a build to run code is annoying.  However, the issue in my mind with using the unbundled files is that it is not the same as what we actually deploy.  And it will limit us from doing some fancier things when using webpack (I currently have an alias to eliminate the if check in the assertion code when assertions are turned off, for instance) because we want the code to run even if it is not "built".

Half the time in the devbuild is jshint and eslint.  So, maybe without that it would be better?

**Comment by albertjan** [comment]
That sounds perfect @rixner. Exactly what I meant. 😄

**Comment by rixner** [comment]
Okay, I'm going to fix ```brun``` to work with my new infrastructure and drop most other commands. 

I will leave a "deprecated" skulpt.py with the "dead" commands so that people can resurrect them and/or ask me to fix them if desired.

Then I may be close to another pull request.  At that point, I think people can start incrementally improving things relatively easily.

**Comment by rixner** [comment]
@bnmnetp do you use ```upload``` and ```doctest``` or are those old and crufty, too?

**Comment by rixner** [comment]
@bnmnetp , @albertjan can one or both of you try [this](/rixner/skulpt/tree/webpack) out?

```npm run brun foo.py```

It should run webpack to build an unoptimized (and un-style-checked) skulpt.js and then open a web browser to serve the files and allow you to run the test.

**Comment by rixner** [comment]
@bnmnetp 

5. I think those tXXX.py files still have value.  I do not think the unit tests are a superset of them.  However, I think we've all agreed to not write any more. :smile: 

6. I'm really not sure what the "best practices" are for this situation.  I figure I will put something out there and people can evolve it towards something better...  I see no need for a monolithic ```skulpt.js``` so I didn't build one...  If others disagree, I don't think it would be that hard to aggregate some things, but there are several commands (like repl) where I don't think it makes sense to gratuitously route them through a unified file just to have everything in one place.


@albertjan 

5. Those will all stick around in skulpt.py then, and we can see what @meredydd wants to do.

6. Yeah, that would be nice.  I don't know if this brings us closer to that.  Maybe?


Otherwise, I think what I have linked above meets all of your requirements.  The only thing potentially missing is the debug flag to brun, but I can add that easily, if you guys think what I have is the right thing.

There probably needs to be a developer readme somewhere to explain how to use everything...  I guess I need to write that, too.

**Comment by bnmnetp** [comment]
@rixner -- I do no use either upload or doctest

If the tXXX.py files still have value then great lets keep them.  I was just looking at a bunch of them and they were really simple tests, like adding two numbers together....  I know there are other more complecated tests than that though also.

Thanks!

**Comment by rixner** [comment]
Current status:

All commands, except the following, have been migrated to Javascript/webpack/node using ```package.json``` scripts.  The following remain in ```skulpt.py```:

1. debugbrowser - this runs the numbered tests in the browser.  It currently does not work.  I'm not sure it makes sense to just run the numbered tests, so I don't know what to do with this one.
2. buildVFS - this is used by debugbrowser
3. browser - this builds the numbered tests in a way that could be used in a browser.  It's not used by any other commands, so I don't know what it's actually for.
4. vmwareregr - this says "todo; not working yet."  I think this was a laudable attempt to do cross-browser testing.  It seems like this should just be deleted.
5. host - this fires up a web server in the local directory.  I don't know why.
6. A bunch of "regen" commands to regenerate tests.
7. upload and doctest - seem to be appengine upload scripts, I don't know if these are legacy or actually used, perhaps they are there for @bnmnetp to easily upload?  If not, I think they should be deleted.

I eliminated the ```time``` command entirely, and just always report run times on individual file runs and test runs.  I also did not ```gzip``` the final distribution file to figure out its size.  Webpack already reports the un-gzipped size.

I propose to delete 1-5 and maybe 7 above with no replacements and leave only 6 in skulpt.py.  Assuming ```brun``` works to everyone's satisfaction, then this is probably ready for a pull request.

I did port ```regenparser``` to node as this is regenerating part of skulpt, not the tests, but it doesn't work.  I filed an issue (#882), as this didn't work before either.

The only two other issues (in my mind) are (1) whether it should force a rebuild (the lighter weight version) every time you run anything or not and (2) how to make sure travis does the right thing (it should now do ```npm run dist```, which hopefully fails in a way that Travis will detect.  Note that currently dist fails because Python 3 tests fail.  Again, they didn't work before, either.  But, we need to have Travis run the Python 3 tests, which we have not been doing.

Another slightly annoying thing is you now have to type ```npm run run foo.py``` to run a file.  The first ```run``` is to tell it to use a ```package.json``` script, the second is our normal ```run``` command.  That trips me up.  There's nothing I can really do about it, I don't think, except maybe to change the second ```run``` to be something else?

**Comment by rixner** [comment]
I guess a key question is whether anyone thinks we should be able to run the entire test suite in the browser, as that's what commands 1-3 in my previous comment were meant to do.  However, I don't think this is a capability we really had before (I don't think you could ever actually do this with the unit tests given the current infrastructure), but it might be a nice one.  I don't know how much work it would be to make that happen.  Or if I should tackle that now or let that be a future task for myself or someone else.

I personally rely on just running the tests in d8/node and not in a browser.  Maybe that's a good idea, maybe it's not.

**Comment by rixner** [comment]
Small update: not gzips the optimized skulpt.min.js, too.  So, webpack will now report the size of both skulpt.min.js and skulpt.min.js.gz.  This adds a little time to the build, but we were doing it before, too.

**Comment by acbart** [comment]
This is a very exciting improvement, especially alongside the change to use webpack. This is a great time for Skulpt!

**Comment by albertjan** [comment]
> All commands, except the following, have been migrated to Javascript/webpack/node using package.json scripts. The following remain in skulpt.py:

This is an awesome achievement, I love that you setup a web server to serve the files. I'll have a look if we can setup express with hot reloading we might have to use a the web pack dev server with a proxy to do that but it should be trivial.

> debugbrowser - this runs the numbered tests in the browser. It currently does not work. I'm not sure it makes sense to just run the numbered tests, so I don't know what to do with this one.
> buildVFS - this is used by debugbrowser
> browser - this builds the numbered tests in a way that could be used in a browser. It's not used by any other commands, so I don't know what it's actually for.
> vmwareregr - this says "todo; not working yet." I think this was a laudable attempt to do cross-browser testing. It seems like this should just be deleted.
> host - this fires up a web server in the local directory. I don't know why.

I think we can do without those.

> A bunch of "regen" commands to regenerate tests.

We'll have to see which of these we'll need in the future.

> upload and doctest - seem to be appengine upload scripts, I don't know if these are legacy or actually used, perhaps they are there for @bnmnetp to easily upload? If not, I think they should be deleted.

If we're moving to the GitHub-pages setup these can probably be deleted.

> I eliminated the time command entirely, and just always report run times on individual file runs and test runs. I also did not gzip the final distribution file to figure out its size. Webpack already reports the un-gzipped size.

sounds good

> I did port regenparser to node as this is regenerating part of skulpt, not the tests, but it doesn't work. I filed an issue (#882), as this didn't work before either.

Fix incoming, I will reveal the py3 stuff I have this week. 😄 

About travis, I hope that I'll have some time to look at that soon. I think it will be quite straight forward as it already calls most scripts through npm.

> Another slightly annoying thing is you now have to type npm run run foo.py to run a file. The first run is to tell it to use a package.json script, the second is our normal run command. That trips me up. There's nothing I can really do about it, I don't think, except maybe to change the second run to be something else?

you could use `npm start` as an alias for run so you can say `npm start foo.py` the start script is a special case just as `test` and `install`.

**Comment by rixner** [comment]
Closing, as I think all of these have been resolved and #889 is ready.
